src/core/qgsmaplayerregistry.cpp

```diff
@@ -203,14 +220,17 @@ void QgsMapLayerRegistry::onMapLayerDeleted( QObject* obj )
 
 QMap<QString, QgsMapLayer*> QgsMapLayerRegistry::mapLayers()
 {
-  return mMapLayers;
+  QReadLocker lock( &mRegistryLock );
+
+  QMap<QString, QgsMapLayer*> mapLayers = mMapLayers;
+  mapLayers.detach();
+  return mapLayers;
```

Comment:
Implicit sharing in Qt uses atomic reference counting, so the detach() call should be safe, but it is not needed in the first place because it would be called automatically when first needed...

http://doc.qt.io/qt-4.8/threads-modules.html#threads-and-implicitly-shared-classes